### PR TITLE
Add support for running BiDi algorithm on text with styling annotations.

### PR DESCRIPTION
**Pull request by ChrisLoer**
- Upgrade to Emscripten 1.38 (requires explicit WASM=0 to build asm.js)
 - Upgrade to ICU 62.1
 - Adds `processStyledBidirectionalText` that takes an array of style indices in parallel to the input text and returns results annotated with the correct (reordered) indices
 - Add basic test for `processStyledBidirectionalText` that that splits contiguous input style sections into discontinguous output sections
 - Regenerate checked-in build outputs.

Note the test strings in arabic.test.js are mind-bending and may break layout in your text viewer (that includes GitHub's preview: those arrays that look broken aren't).

The previous approach (which we still use for unstyled text) was:
 - Run BiDi on the whole paragaph
 - Determine line breaks
 - Run ICU's `setLine` with chosen line break points to prepare individual lines
 - Run ICU's `writeReordered` to correctly transform/reorder all logical runs within the line into visual runs.

The new approach for styled lines is:
 - Run BiDi on the whole paragaph
 - Determine line breaks
 - Run ICU's `setLine` with chosen line break points to prepare individual lines
 - Use ICU's `getVisualRun` to iterate over directional runs within a line in visual order. For each run:
   - if it's LTR, just copy it into the output, no transformations are needed
   - if it's RTL, iterate over it _logically backwards_ (visually forwards), breaking into segments by style. For each segment, use ICU's `writeReverse` to reverse the section _and_ apply transformations. This may change the length of the segment, but since the whole segment shares the same style, we can just apply that style to every character in whatever the reversed output is. Note that `writeReverse` is responsible for keeping combining characters together in the right (aka still "logical") order. If you changed style mid-combining character, this algorithm could move the combining character to the wrong place, and you would deserve whatever happened to you.

One open question in my mind is whether the style annotation interface makes sense (an array of annotations maintained in parallel with a flat string). Alternatively, we could represent the input in terms of sections (aka instead of representing a section "foo" with style 0 as `["foo", [0, 0, 0]]`, we could represent it as `["foo",0]`). Internally, it was easier for me to work with a data structure that made it super-easy to go from "source string logical index" -> "style", but it'd be easy enough to do the conversion. The thing is, I'm not sure an annotated-per-section interface is actually easier to work with, because the caller then has to keep track of iterating over sections too, and it's confusing that a contiguous input section can actually be split up into multiple discontiguous output sections...

/cc @anandthakker

### CODE REVIEW

**Comment by ChrisLoer** [comment]
Oh also this is meant to be backwards compatible with gl-js in both directions. Old versions that load this should just ignore `processStyledBidirectionalText`. New versions of gl-js that load an old version of the rtl-text plugin will still be able to run BiDi for unstyled text, but won't run it for styled text (see `shaping.js` logic at https://github.com/mapbox/mapbox-gl-js/pull/6994).

File: src/icu.js [hunk 1]
```diff
@@ -80,3 +90,94 @@ function processBidirectionalText(input, lineBreakPoints) {
 
     return lines;
 }
+
+function createInt32Ptr() {
+    return Module._malloc(4);
+}
+
+function readInt32Ptr(ptr) {
```
Comment (anandthakker) [inline src/icu.js]: Would it make sense to call this something like `readAndFreeInt32Ptr` or `consumeInt32Ptr`?

File: src/icu.js [hunk 2]
```diff
@@ -80,3 +90,94 @@ function processBidirectionalText(input, lineBreakPoints) {
...
+
+    if (returnStringPtr === 0) {
+        return null;
+    }
+    const reversed = Module.UTF16ToString(returnStringPtr);
+    Module._free(returnStringPtr);
+    return reversed;
+}
+
+function processStyledBidirectionalText(text, styleIndices, lineBreakPoints) {
```
Comment (anandthakker) [inline src/icu.js]: Add brief comment describing what `styleIndices` should be and what the return value looks like

File: src/icu.js [hunk 3]
```diff
@@ -80,3 +90,94 @@ function processBidirectionalText(input, lineBreakPoints) {
...
+    const nDataBytes = (text.length + 1) * 2;
+    const stringInputPtr = Module._malloc(nDataBytes);
+    const paragraphCount = setParagraph(text, stringInputPtr, nDataBytes);
+    if (!paragraphCount) {
+        return [{text, styleIndices}];
+    }
+
+    const mergedParagraphLineBreakPoints = mergeParagraphLineBreakPoints(lineBreakPoints, paragraphCount);
+
+    let startIndex = 0;
```
Comment (anandthakker) [inline src/icu.js]: Maybe `lineStartIndex`?

**Review by anandthakker — COMMENTED** [review]
Looks good overall to me!

> The thing is, I'm not sure an annotated-per-section interface is actually easier to work with, because the caller then has to keep track of iterating over sections too, and it's confusing that a contiguous input section can actually be split up into multiple discontiguous output sections

Yeah, I agree that the per-section approach doesn't sound particularly easier to work with.  It would presumably take less space, though, right?  Not sure if the difference would be enough to matter...

**Comment by ChrisLoer** [comment]
> It would presumably take less space, though, right? Not sure if the difference would be enough to matter...

Yeah, I'm going to start with the assumption that it doesn't matter. We're not holding onto these arrays, so the cost would just show up... I dunno, maybe with the wasm version of mapbox-gl-rtl-text the overhead of transferring the array would be greater? But it seems reasonable to ignore unless it actually shows up as significant in profiling.

**Comment by ChrisLoer** [comment]
Huh, ICU's download site appears to be down, which breaks the build. I'll try again later.

**Review by anandthakker — APPROVED** [review]

**Pull request merged**